### ISSUE TITLE
Producers and Consumers should reconnect transparently

### ISSUE THREAD

**Issue by stearnsc** [streamnative/pulsar-rs#28]
Right now, each producer or consumer has a single connection, and if that connection disconnects, the producer or consumer must be recreated. It would be nice to make this happen transparently, possibly with a configurable timeout.

Some issues that need to be resolved: 

(1) What's the best abstraction for the producers / consumers to use in creating a new connection? Each could contain a full `Pulsar` client and use that to reconnect (easy to accomplish if we limit creating to being done via the client). This has the benefit of allowing better connection re-use (if each client references the same connection manager, they can get references to existing connections without difficulty).

(2) How do we handle enqueued messages? Currently the producer and the connection tasks communicate using futures::sync::mpsc channels; in the event of a reconnect, any messages enqueued would be dropped. Is there a clean way to recover them and re-send? (Could have a "return messages" queue or something that gets filled with any pending messages in a `Drop` impl, but this starts adding complexity that makes me nervous).

(3) The machinery around swapping out a connection is nontrivial in the producer itself; reconnecting will resolve with a `Future` of `Connection`, but will need some ability to actually mutate the consumer or producer, which implies some sort of sync mutability, but I don't really like the idea of every connection being behind a lock & refcell, so possibly there are some architecture improvements that could be made there to make that easier?

**Comment by Geal** [comment]
reconnection was done in #95.
Producers and consumers do as (1), holding a `Pulsar` instance to create new connection, but service discovery calls directly the connection manager.
Point (2) is not done yet but should not be too hard (see #100)
Point (3) was solved by making the producer mutable (it does not change usage more than writing `let mut producer` instead of `let producer`). The consumer is not mutable because its behaviour is hidden in a spawned task (the only way I found to implement `Stream` with async/await)
